File: 프로그래머스/unrated/138476. 귤 고르기/귤 고르기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <map>
using namespace std;

int solution(int k, vector<int> tangerine) {
    int answer = 0;
    map<int, int> m; //귤 크기별 개수 [크기, 개수]
    
    for(int i=0; i<tangerine.size(); i++){
        m[tangerine[i]]++;
    }
    
    vector<int> v;
    map<int, int>::iterator iter; 
    for(iter = m.begin(); iter!=m.end(); iter++){
        v.push_back((*iter).second); 
    }
    sort(v.begin(), v.end(), greater<int>());
    
    int cnt = 0; 
    for(int i=0; i<v.size(); i++){
        if(cnt>=k) break;
        answer++; //종류 수
        cnt+=v[i]; //담은 귤 수 
    }
    
    return answer;
}

File: 프로그래머스/2/138476. 귤 고르기/귤 고르기.java
import java.util.*;

class Solution {
    public int solution(int k, int[] tangerine) {
        int answer = 0;
        
        Arrays.sort(tangerine);
        int len = tangerine.length;
        int maxSize = tangerine[len-1];
        
        int[] arr = new int[maxSize+1];
        
        for(int size : tangerine){
            arr[size]++;
        }
        
        Arrays.sort(arr);
        
        int sum = 0; //현재까지 담은 귤 수
        for(int i=maxSize; i>0; i--){
            if(arr[i]==0){
                continue;
            }
            if((sum+arr[i])<k){
                sum+= arr[i];
                answer++;
            }else{
                answer++;
                break;
            }
        }
        return answer;
    }
}